samza-core/src/test/java/org/apache/samza/zk/TestZkBarrierForVersionUpgrade.java

```diff
@@ -18,205 +18,221 @@
-        barrierId,
-        zkUtils,
-        new ZkBarrierListener() {
-          @Override
-          public void onBarrierCreated(String version) {
-          }
-
-          @Override
-          public void onBarrierStateChanged(String version, ZkBarrierForVersionUpgrade.State state) {
-            if (ZkBarrierForVersionUpgrade.State.TIMED_OUT.equals(state)) {
-              timeoutStateChangeCalled.incrementAndGet();
-              latch.countDown();
-            }
-          }
-
-          @Override
-          public void onBarrierError(String version, Throwable t) {
-
-          }
-
-        });
-    processor1Barrier.create(ver, processors);
-    processor1Barrier.join(ver, "p1");
-
-    final ZkBarrierForVersionUpgrade processor2Barrier = new ZkBarrierForVersionUpgrade(
-        barrierId,
-        zkUtils1,
-        new ZkBarrierListener() {
-          @Override
-          public void onBarrierCreated(String version) {
-          }
-
-          @Override
-          public void onBarrierStateChanged(String version, ZkBarrierForVersionUpgrade.State state) {
-            if (ZkBarrierForVersionUpgrade.State.TIMED_OUT.equals(state)) {
-              timeoutStateChangeCalled.incrementAndGet();
-              latch.countDown();
-            }
-          }
-
-          @Override
-          public void onBarrierError(String version, Throwable t) {
-
-          }
-
-        });
-
-    processor2Barrier.join(ver, "p2");
-
-    processor1Barrier.expire(ver);
+  @Test
+  public void testShouldDiscardBarrierUpdateEventsAfterABarrierIsMarkedAsDone() {
+    String barrierId = String.format("%s/%s", zkUtils1.getKeyBuilder().getRootPath(), RandomStringUtils.randomAlphabetic(4));
+    List<String> processors = ImmutableList.of("p1", "p2");
+
+    CountDownLatch latch = new CountDownLatch(2);
+    AtomicInteger stateChangedCalled = new AtomicInteger(0);
+    TestZkBarrierListener listener = new TestZkBarrierListener(latch, stateChangedCalled, State.DONE);
+    ZkBarrierForVersionUpgrade processor1Barrier = new ZkBarrierForVersionUpgrade(barrierId, zkUtils, listener);
+    ZkBarrierForVersionUpgrade processor2Barrier = new ZkBarrierForVersionUpgrade(barrierId, zkUtils1, listener);
+
+    processor1Barrier.create(BARRIER_VERSION, processors);
+
+    processor1Barrier.join(BARRIER_VERSION, "p1");
+    processor2Barrier.join(BARRIER_VERSION, "p2");
+
     boolean result = false;
     try {
       result = latch.await(10000, TimeUnit.MILLISECONDS);
     } catch (InterruptedException e) {
       e.printStackTrace();
     }
-    Assert.assertTrue("Barrier Timeout test failed to complete within test timeout.", result);
+    assertTrue("Barrier Timeout test failed to complete within test timeout.", result);
+
+    processor1Barrier.expire(BARRIER_VERSION);
+
+    State barrierState = zkUtils.getZkClient().readData(barrierId + "/barrier_1/barrier_state");
+    assertEquals(State.DONE, barrierState);
+  }
+
+  @Test
```

Comment:
Should we add tests to ensure that DONE->TIMED_OUT and TIMED_OUT -> DONE have no effect?